### DISCUSSION TITLE
Upcoming features for consumers of realtime/historical WZDx data

### DISCUSSION THREAD

**Discussion by chuehlien** [usdot-jpo-ode/jpo-wzdx#49] [General]
As @arielsgold had mentioned in https://github.com/usdot-jpo-ode/jpo-wzdx/issues/37, the USDOT ITS DataHub team is in the process of creating a pipeline to aggregate the WZDx data feeds as a proof of concept. We see two potential user groups - those that hope to consume the data in realtime and those that hope to perform analytics on the historical data. Hoping to get some feedback on the following to see what you would find most useful and get any additional feedback:

**Realtime Data Consumer** (e.g. mapping companies)
At the moment, we think it's best that the realtime data consumer hit the WZDx feed directly. However, we could provide the following resources:
- **Directory of WZDx feeds**: We'd like to maintain a directory of active WZDx feeds, as @barbeau, @davidcraig4300, and @DeraldDudley have described. The directory of feeds will also include information on key metadata such as URL, WZDx schema version, update frequency, and perhaps data veracity, as @DeLongR, @sknick-iastate, and @orbonjb have flagged. For the initial phase, we would likely maintain this data in an integrated Socrata dataset that can be queried via an API. Down the line, we'd love to look into having something more sophisticated like https://transit.land/ that allows for user inputs, or look into the webhook "push on change" type of capability that @bdelsey64 has mentioned, at a feed directory level. Also related: https://github.com/usdot-jpo-ode/jpo-wzdx/issues/37, https://github.com/usdot-jpo-ode/jpo-wzdx/issues/18.
- **Data wrangling toolkit**: We'd also provide importable Python package on GitHub for transforming data from one WZDx format (JSON, XML, GeoJSON) and schema version to some other output format (CSV or WZDx JSON/XML/GeoJSON), since we'll have written this during our data archiving exercise.

**Historical Data Consumer** (e.g. government agencies, ITS researchers)
We plan to ingest and archive the existing WZDx data feeds into an AWS S3 bucket, organized using the following naming convention in a JSON newline file:
`/{State}/{issuingOrganization}/{Year}/{Month}/{identifier}_{BeginLocation.RoadDirection}_{YYYYMM}.txt`

This means that each file would contain the statuses of one work zone over the span of the month. 

Each WZDx feed will be polled based on the feed's update frequency. Since we don't foresee a work zone's status to change much throughout the day, we thought it might make sense to only save a status when it's different from the previous status to minimize repeated data. We see four potential approaches for reducing duplicate data:

- Option 1: Save, at a minimum, the first and most recent status of the work zone for the month. Also save statuses that "bookend" any detected change (e.g. when the current status is different from the previous status, save both the current and previous status in the archive.
- Option 2: Save, at a minimum, the first and most recent status of the work zone for the month AND the last status of each day so that it is explicit that the data archive has been updated at least daily even when the work zone status remains the same for a whole month. Also save statuses that bookend any changes.
- Option 3: Save all statuses retrieved for the current month. At the end of the month, clean up duplicate statuses to make the file look like what's described in Option 1 and 2.
- Option 4: Save all statuses retrieved.

As our goal is to create something that is useful to the data consumers and to inspire more data providers to share data via the open standard, your feedback is most valuable to us! Your feedback will shape the work we do - additional comments are also appreciated.

Julia Lien, USDOT ITS DataHub

**Comment by barbeau** [comment]
From my experience from working with archiving GTFS-realtime data for transit, I would strongly suggest archiving the raw WZDx feed content prior to parsing it - this sounds like Option 3 or 4. Sometimes you'll end up with data that your archiving software isn't designed to interpret (e.g., different semantic interpretation of fields), or new custom fields that your parser doesn't pick up. Having the raw data feed to go back and review the canonical source data has proven very useful. There are lossless ways of compressing this data that can help save space.

Also, I'd make the assumption that some requests to WXDx feeds won't return valid data - it could be malformed, be an empty feed, an HTTP 404, etc. I'd suggest finding a way to represent this information in the archive so you can also tell where data might be missing.

Related to this - will this system have some kind of real-time monitoring that will alert someone when archiving gets interrupted for a particular feed? A way to alert producers when feeds stop working is a common problem that GTFS, GTFS-realtime, and GBFS have all run into.

**Comment by chuehlien** [comment]
Thank you for sharing your invaluable feedback @barbeau!

1. Archiving raw WZDx feed content prior to parsing: 
Great suggestion to add resiliency to the archiving mechanism. Based on your experience, would you recommend that both the raw content (snapshots of whatever a feed returns) and the semi-parsed content (statuses separated out by work zone identifier) be available to users? Or would you recommend that the raw content be saved only temporarily at a "staging" database and be deleted as data gets parsed into the semi-processed format successfully?

2. Finding ways to represent malformed or empty feed data in the archive.
Great point. We'll look into this! 

3. Monitoring of the archival process that provides feedback to the producers.
We don't have this in place with our current MVP implementation, but we'd love to prioritize this in our next iteration. Would you recommend that the producer receives an alert and that we host a status page showing realtime status of all feeds?

Could we look for a time to talk through your feedback? I have additional questions around 2, 3, and how the archive could deal with evolving schema. Would love to get your thoughts given your experience with archiving GTFS feed.

**Comment by barbeau** [comment]
>Based on your experience, would you recommend that both the raw content (snapshots of whatever a feed returns) and the semi-parsed content (statuses separated out by work zone identifier) be available to users?

I'd definitely recommend keeping the raw content snapshots in a permanent archive so you always have a source of truth to reference when someone questions the semi-parsed content. I'm not sure if I have a good enough handle on the various users of this archive to guess whether the raw archive or semi-parsed content would be more useful, but if you're going to keep a permanent archive of the raw snapshots anyway you might as well make if available to users in some manner (e.g., zipped archive).

>Would you recommend that the producer receives an alert and that we host a status page showing realtime status of all feeds?

I think both would be great. Keeping contact info up-to-date for feeds is always a challenge, but maybe for producers that don't have internal monitoring set up this would actually be an incentive to keep their contact info updated.

>Could we look for a time to talk through your feedback?

Sure, send me an invite via LinkedIn and we can schedule a time to talk - https://www.linkedin.com/in/seanbarbeau/.
